File: adapters/php/swapbook.php
<?php
// Swapbook adapter for Laravel / PHP.
//
// Exposes the Swapbook protocol (manifest / preview / mocks / mock). Render-
// agnostic: a variant's render is a callable(array $args): string returning
// HTML, so it works with Blade, plain PHP, or any templating. Dispatch requests
// through Swapbook::handle(); mount point is /_swapbook.

function sb_control(string $name, string $type = 'text', $default = null, ?array $options = null): array
{
    $c = ['name' => $name, 'type' => $type, 'default' => $default];
    if ($options !== null) {
        $c['options'] = $options;
    }
    return $c;
}

function sb_mock(string $route, callable $render): array
{
    [$verb, $path] = str_contains($route, ' ') ? explode(' ', $route, 2) : ['GET', $route];
    return ['verb' => strtoupper($verb), 'path' => $path, 'render' => $render];
}

function sb_variant(string $name, callable $render, array $controls = [], string $docs = '', array $mocks = []): array
{
    return compact('name', 'render', 'controls', 'docs', 'mocks');
}

class Swapbook
{
    private array $stories = [];
    public string $htmxSrc = '';
    public string $cssSrc = '';
    public string $jsSrc = '';

    public function register(string $name, array $variants, string $group = '', string $docs = ''): void
    {
        $this->stories[] = ['id' => self::slug($name), 'name' => $name, 'group' => $group, 'docs' => $docs, 'variants' => $variants];
    }

    public static function slug(string $s): string
    {
        return trim(preg_replace('/-+/', '-', preg_replace('/[^a-z0-9]+/', '-', strtolower($s))), '-');
    }

    private function find(string $id, string $vname): ?array
    {
        foreach ($this->stories as $s) {
            if ($s['id'] === $id) {
                foreach ($s['variants'] as $v) {
                    if ($v['name'] === $vname) {
                        return $v;
                    }
                }
            }
        }
        return null;
    }

    private function coerce(array $controls, array $q): array
    {
        $args = [];
        foreach ($controls as $c) {
            $n = $c['name'];
            if (!array_key_exists($n, $q)) { // absent -> default; empty is a real value
                $args[$n] = $c['default'] ?? null;
                continue;
            }
            $raw = (string) $q[$n];
            $args[$n] = match ($c['type']) {
                'number' => is_numeric($raw) ? (float) $raw : ($c['default'] ?? 0),
                'bool' => in_array($raw, ['true', '1', 'on'], true),
                default => $raw,
            };
        }
        return $args;
    }

    /** @return array{0:int,1:string,2:string} [status, contentType, body] */
    public function handle(string $method, string $path, array $query): array
    {
        if ($path === '/_swapbook/manifest.json') {
            return $this->json($this->manifest());
        }
        if (preg_match('#^/_swapbook/preview/([^/]+)/([^/]+)$#', $path, $m)) {
            $v = $this->find($m[1], $m[2]);
            return $v ? $this->html(($v['render'])($this->coerce($v['controls'], $query))) : $this->notFound();
        }
        if (preg_match('#^/_swapbook/mocks/([^/]+)/([^/]+)$#', $path, $m)) {
            $v = $this->find($m[1], $m[2]);
            if (!$v) {
                return $this->notFound();
            }
            $out = [];
            foreach ($v['mocks'] as $i => $mk) {
                $out[] = ['verb' => $mk['verb'], 'path' => $mk['path'], 'index' => $i];
            }
            return $this->json($out);
        }
        if (preg_match('#^/_swapbook/mock/([^/]+)/([^/]+)/(\d+)$#', $path, $m)) {
            $v = $this->find($m[1], $m[2]);
            return ($v && isset($v['mocks'][(int) $m[3]]))
                ? $this->html(($v['mocks'][(int) $m[3]]['render'])([]))
                : $this->notFound();
        }
        return $this->notFound();
    }

    private function manifest(): array
    {
        return [
            'htmxSrc' => $this->htmxSrc, 'cssSrc' => $this->cssSrc, 'jsSrc' => $this->jsSrc,
            'stories' => array_map(fn($s) => [
                'id' => $s['id'], 'name' => $s['name'], 'group' => $s['group'], 'docs' => $s['docs'],
                'variants' => array_map(fn($v) => ['name' => $v['name'], 'controls' => $v['controls'], 'docs' => $v['docs']], $s['variants']),
            ], $this->stories),
        ];
    }

    private function json(array $o): array { return [200, 'application/json', json_encode($o)]; }
    private function html(string $s): array { return [200, 'text/html; charset=utf-8', $s]; }
    private function notFound(): array { return [404, 'text/plain', 'not found']; }
}

File: adapters/rails/swapbook.rb
# Swapbook adapter for Rails (and any Rack app).
#
# A Rack app exposing the Swapbook protocol (manifest / preview / mocks / mock).
# Render-agnostic: a variant's :render is a proc (args) -> HTML string, so it
# works with ActionView partials, ViewComponent, Phlex or raw strings, on any
# modern Rails (6+). Mount it:  mount registry => "/_swapbook"
require "json"
require "rack"

module Swapbook
  def self.slug(s)
    s.downcase.gsub(/[^a-z0-9]+/, "-").gsub(/-+/, "-").gsub(/\A-|-\z/, "")
  end

  def self.control(name, type: "text", default: nil, options: nil)
    { name: name, type: type, default: default, options: options }.compact
  end

  # Pass a non-2xx status (422, 500, …) to preview a component's error state.
  def self.mock(route, render, status: 200)
    verb, path = route.include?(" ") ? route.split(" ", 2) : ["GET", route]
    { verb: verb.upcase, path: path, render: render, status: status }
  end

  def self.variant(name, render, controls: [], docs: "", mocks: [])
    { name: name, render: render, controls: controls, docs: docs, mocks: mocks }
  end

  class Registry
    def initialize(htmx_src: "", css_src: "", js_src: "")
      @htmx_src, @css_src, @js_src = htmx_src, css_src, js_src
      @stories = []
      @global_mocks = []
    end

    def register(name, variants:, group: "", docs: "")
      @stories << { id: Swapbook.slug(name), name: name, group: group, docs: docs, variants: variants }
    end

    # Declare a registry-level mock merged into every variant, for routes shared
    # across stories. A variant's own mock for the same route wins. Chainable.
    def mock(route, render, status: 200)
      @global_mocks << Swapbook.mock(route, render, status: status)
      self
    end

    # Rack entry point (PATH_INFO is relative to the mount point).
    def call(env)
      path = env["PATH_INFO"]
      params = Rack::Request.new(env).params
      case path
      when "/manifest.json"
        json(manifest)
      when %r{\A/preview/([^/]+)/([^/]+)\z}
        v = find($1, $2)
        v ? html(v[:render].call(coerce(v[:controls], params))) : not_found
      when %r{\A/mocks/([^/]+)/([^/]+)\z}
        v = find($1, $2)
        v ? json(mocks_for(v).each_with_index.map { |m, i| { verb: m[:verb], path: m[:path], index: i } }) : not_found
      when %r{\A/mock/([^/]+)/([^/]+)/(\d+)\z}
        v = find($1, $2)
        mk = v && mocks_for(v)[$3.to_i]
        mk ? html(mk[:render].call({}), mk[:status] || 200) : not_found
      else
        not_found
      end
    end

    private

    def manifest
      {
        htmxSrc: @htmx_src, cssSrc: @css_src, jsSrc: @js_src,
        stories: @stories.map { |s|
          {
            id: s[:id], name: s[:name], group: s[:group], docs: s[:docs],
            variants: s[:variants].map { |v| { name: v[:name], controls: v[:controls], docs: v[:docs] } },
          }
        },
      }
    end

    def find(sid, vname)
      s = @stories.find { |st| st[:id] == sid }
      s && s[:variants].find { |v| v[:name] == vname }
    end

    # registry-level mocks first, then the variant's own (which override on dup)
    def mocks_for(v)
      @global_mocks + v[:mocks]
    end

    def coerce(controls, params)
      controls.each_with_object({}) do |c, args|
        name = c[:name]
        unless params.key?(name) # absent -> default; present-but-empty is real
          args[name] = c[:default]
          next
        end
        raw = params[name].to_s
        args[name] =
          case c[:type]
          when "number" then (Float(raw) rescue c[:default])
          when "bool" then %w[true 1 on].include?(raw)
          else raw
          end
      end
    end

    def json(obj) = [200, { "content-type" => "application/json" }, [JSON.generate(obj)]]
    def html(str, status = 200) = [status, { "content-type" => "text/html; charset=utf-8" }, [str.to_s]]
    def not_found = [404, { "content-type" => "text/plain" }, ["not found"]]
  end
end
